File: src/services/openai/unifiedClient.ts
/**
 * Unified OpenAI Client Wrapper
 * 
 * Provides consistent client initialization, credential resolution, and health checks
 * following OpenAI SDK best practices and Railway-native patterns.
 * 
 * Features:
 * - Railway-first credential resolution with fallbacks
 * - Stateless, deterministic initialization
 * - Health check capabilities
 * - Type-safe model selection
 * - Audit trail for all operations
 * 
 * @module unifiedClient
 */

import OpenAI from 'openai';
import { aiLogger } from '../../utils/structuredLogging.js';
import { recordTraceEvent } from '../../utils/telemetry.js';
import {
  resolveOpenAIKey,
  resolveOpenAIBaseURL,
  getOpenAIKeySource,
  hasValidAPIKey,
  setDefaultModel,
  getDefaultModel,
  getFallbackModel,
  getGPT5Model,
  getComplexModel
} from './credentialProvider.js';
import {
  getCircuitBreakerSnapshot,
  RESILIENCE_CONSTANTS
} from './resilience.js';
import { responseCache } from '../../utils/cache.js';
import { getRoutingActiveMessage } from '../../config/prompts.js';
import { getConfig } from '../../config/unifiedConfig.js';
import { APPLICATION_CONSTANTS } from '../../utils/constants.js';
import { resolveErrorMessage } from '../../lib/errors/index.js';

/**
 * Client initialization options
 */
export interface ClientOptions {
  /** API key override (defaults to environment resolution) */
  apiKey?: string;
  /** Base URL override (defaults to environment resolution) */
  baseURL?: string;
  /** Timeout in milliseconds (defaults to WORKER_API_TIMEOUT_MS or 60000) */
  timeout?: number;
  /** Whether to use singleton pattern (default: true) */
  singleton?: boolean;
}

/**
 * Health status for OpenAI client
 */
export interface HealthStatus {
  /** Whether client is initialized and healthy */
  healthy: boolean;
  /** Whether API key is configured */
  apiKeyConfigured: boolean;
  /** Source of API key */
  apiKeySource: string | null;
  /** Default model configured */
  defaultModel: string;
  /** Fallback model configured */
  fallbackModel: string;
  /** Circuit breaker state */
  circuitBreakerHealthy: boolean;
  /** Cache statistics */
  cacheEnabled: boolean;
  /** Last health check timestamp */
  lastCheck: string;
  /** Error message if unhealthy */
  error?: string;
}

/**
 * Singleton client instance
 */
let singletonClient: OpenAI | null = null;
let initializationAttempted = false;

/**
 * API timeout from config or default
 */
export const API_TIMEOUT_MS = (() => {
  const config = getConfig();
  return config.workerApiTimeoutMs;
})();

/**
 * ARCANOS routing message for all completions
 */
export const ARCANOS_ROUTING_MESSAGE = getRoutingActiveMessage();

/**
 * Creates a new OpenAI client with Railway-native credential resolution
 * 
 * This function follows Railway best practices:
 * - Stateless initialization (no local state dependencies)
 * - Environment variable resolution with Railway fallbacks
 * - Deterministic behavior (same inputs = same outputs)
 * - Comprehensive error handling and logging
 * 
 * @param options - Client initialization options
 * @returns OpenAI client instance or null if initialization fails
 */
export function createOpenAIClient(options: ClientOptions = {}): OpenAI | null {
  const startTime = Date.now();
  const traceId = recordTraceEvent('openai.client.create.start', {
    hasApiKeyOverride: Boolean(options.apiKey),
    hasBaseURLOverride: Boolean(options.baseURL),
    timeout: options.timeout || API_TIMEOUT_MS
  });

  try {
    // Resolve API key with Railway fallbacks
    const apiKey = options.apiKey || resolveOpenAIKey();
    
    //audit Assumption: missing API key means OpenAI calls must be mocked; risk: real calls without key fail; invariant: return null without key; handling: log and short-circuit.
    if (!apiKey) {
      aiLogger.warn('OpenAI API key not configured - AI endpoints will return mock responses', {
        operation: 'createOpenAIClient',
        module: 'openai.unified'
      });
      recordTraceEvent('openai.client.create.no_key', { traceId });
      return null;
    }

    // Resolve base URL with Railway fallbacks
    const baseURL = options.baseURL || resolveOpenAIBaseURL();
    const timeout = options.timeout || API_TIMEOUT_MS;
    const config = getConfig();

    // Create client instance
    const client = new OpenAI({
      apiKey,
      timeout,
      maxRetries: config.openaiMaxRetries,
      ...(baseURL ? { baseURL } : {})
    });

    // Configure default model from config (adapter boundary pattern)
    const configuredDefaultModel = config.defaultModel || APPLICATION_CONSTANTS.MODEL_GPT_4O_MINI;
    setDefaultModel(configuredDefaultModel);

    const duration = Date.now() - startTime;
    aiLogger.info('✅ OpenAI client created', {
      module: 'openai.unified',
      operation: 'createOpenAIClient',
      duration,
      model: configuredDefaultModel,
      source: getOpenAIKeySource()
    });

    recordTraceEvent('openai.client.create.success', {
      traceId,
      duration,
      model: configuredDefaultModel
    });

    return client;
  } catch (error) {
    const duration = Date.now() - startTime;
    const errorMessage = resolveErrorMessage(error);
    
    aiLogger.error('❌ Failed to create OpenAI client', {
      module: 'openai.unified',
      operation: 'createOpenAIClient',
      duration
    }, undefined, error as Error);

    recordTraceEvent('openai.client.create.error', {
      traceId,
      duration,
      error: errorMessage
    });

    return null;
  }
}

/**
 * Gets or creates the singleton OpenAI client
 * 
 * Uses singleton pattern for consistent client reuse across the application.
 * Initializes on first call with Railway-native credential resolution.
 * 
 * @returns OpenAI client instance or null if initialization fails
 */
export function getOrCreateClient(): OpenAI | null {
  if (singletonClient) {
    return singletonClient;
  }

  // If we previously failed (no key) but a key is now available, allow one retry (e.g. env loaded late or set in platform)
  if (initializationAttempted && hasValidAPIKey()) {
    aiLogger.info('OpenAI API key now available - retrying client creation', {
      module: 'openai.unified',
      operation: 'getOrCreateClient'
    });
    initializationAttempted = false;
    singletonClient = null;
  }

  if (initializationAttempted) {
    aiLogger.warn('OpenAI client initialization already attempted, returning null', {
      module: 'openai.unified',
      operation: 'getOrCreateClient'
    });
    return null;
  }

  initializationAttempted = true;
  singletonClient = createOpenAIClient({ singleton: true });

  return singletonClient;
}

/**
 * Validates OpenAI client health
 * 
 * Performs comprehensive health check including:
 * - API key configuration
 * - Client initialization status
 * - Circuit breaker state
 * - Cache status
 * 
 * @returns Health status object
 */
export function validateClientHealth(): HealthStatus {
  const circuitBreakerMetrics = getCircuitBreakerSnapshot();
  const configured = hasValidAPIKey();
  const initialized = singletonClient !== null;

  const health: HealthStatus = {
    healthy: configured && initialized && circuitBreakerMetrics.state !== 'OPEN',
    apiKeyConfigured: configured,
    apiKeySource: getOpenAIKeySource(),
    defaultModel: getDefaultModel(),
    fallbackModel: getFallbackModel(),
    circuitBreakerHealthy: circuitBreakerMetrics.state !== 'OPEN',
    cacheEnabled: true,
    lastCheck: new Date().toISOString()
  };

  if (!health.healthy) {
    if (!configured) {
      health.error = 'API key not configured';
    } else if (!initialized) {
      health.error = 'Client not initialized';
    } else if (!health.circuitBreakerHealthy) {
      health.error = 'Circuit breaker is OPEN';
    }
  }

  return health;
}

/**
 * Resets the singleton client
 * 
 * Useful for testing or when credentials change.
 * Clears singleton instance and allows re-initialization.
 */
export function resetClient(): void {
  singletonClient = null;
  initializationAttempted = false;
  recordTraceEvent('openai.client.reset', {
    module: 'openai.unified'
  });
  aiLogger.info('OpenAI client reset', {
    module: 'openai.unified',
    operation: 'resetClient'
  });
}

/**
 * Gets the current singleton client without creating a new one
 * 
 * @returns Current singleton client or null if not initialized
 */
export function getClient(): OpenAI | null {
  return singletonClient;
}

/**
 * Exports for backward compatibility with existing code
 */
export {
  getDefaultModel,
  getFallbackModel,
  getComplexModel,
  getGPT5Model,
  hasValidAPIKey,
  getOpenAIKeySource,
  resolveOpenAIKey,
  resolveOpenAIBaseURL
};

/**
 * Default export for convenience
 */
export default {
  createOpenAIClient,
  getOrCreateClient,
  getClient,
  validateClientHealth,
  resetClient,
  getDefaultModel,
  getFallbackModel,
  getComplexModel,
  getGPT5Model,
  hasValidAPIKey,
  API_TIMEOUT_MS,
  ARCANOS_ROUTING_MESSAGE
};

File: daemon-python/arcanos/openai/unified_client.py
"""
Unified OpenAI Client Wrapper for Python CLI Agent

Provides consistent client initialization, credential resolution, and health checks
following OpenAI SDK best practices and Railway-native patterns.

Features:
- Railway-first credential resolution with fallbacks
- Stateless, deterministic initialization
- Health check capabilities
- Type-safe model selection
- Audit trail for all operations
"""

from typing import Optional, Dict, Any
from datetime import datetime
from openai import OpenAI
from ..config import Config
from ..utils.telemetry import record_trace_event
import logging

logger = logging.getLogger("arcanos.openai")

# Singleton client instance
_singleton_client: Optional[OpenAI] = None
_initialization_attempted = False

# API timeout default (config should be injected)
API_TIMEOUT_MS_DEFAULT = 60000
_OPENAI_KEY_PLACEHOLDERS = {"", "your-openai-api-key-here", "your-openai-key-here"}


def _normalize_secret_candidate(value: Optional[str]) -> Optional[str]:
    """
    Purpose: Normalize and validate secret-like string values.
    Inputs/Outputs: optional raw value; returns trimmed valid value or None.
    Edge cases: Placeholder values are treated as missing.
    """
    if not value:
        return None
    normalized = value.strip()
    if not normalized:
        return None
    if normalized.lower() in _OPENAI_KEY_PLACEHOLDERS:
        return None
    return normalized


class ClientOptions:
    """Client initialization options"""
    def __init__(
        self,
        api_key: Optional[str] = None,
        base_url: Optional[str] = None,
        timeout: Optional[int] = None,
        singleton: bool = True
    ):
        self.api_key = api_key
        self.base_url = base_url
        self.timeout = timeout or API_TIMEOUT_MS_DEFAULT
        self.singleton = singleton


class HealthStatus:
    """Health status for OpenAI client"""
    def __init__(
        self,
        healthy: bool,
        api_key_configured: bool,
        api_key_source: Optional[str],
        default_model: str,
        fallback_model: str,
        error: Optional[str] = None
    ):
        self.healthy = healthy
        self.api_key_configured = api_key_configured
        self.api_key_source = api_key_source
        self.default_model = default_model
        self.fallback_model = fallback_model
        self.error = error
        self.last_check = None

    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary for JSON serialization"""
        return {
            "healthy": self.healthy,
            "apiKeyConfigured": self.api_key_configured,
            "apiKeySource": self.api_key_source,
            "defaultModel": self.default_model,
            "fallbackModel": self.fallback_model,
            "error": self.error,
            "lastCheck": self.last_check
        }


def resolve_openai_key(config: Optional[Config] = None) -> Optional[str]:
    """
    Resolves OpenAI API key from Config.
    
    Args:
        config: Config instance (preferred). If None, uses module-level Config.
    
    Returns:
        API key string or None if not found
    """
    config_to_use = config or Config
    configured_key = _normalize_secret_candidate(getattr(config_to_use, "OPENAI_API_KEY", None))
    #audit Assumption: Config is canonical for runtime credentials; risk: hidden env drift; invariant: Config value preferred when present; handling: return normalized configured key.
    if configured_key:
        return configured_key
    return None


def resolve_openai_base_url(config: Optional[Config] = None) -> Optional[str]:
    """
    Resolves OpenAI base URL from Config.
    
    Args:
        config: Config instance (preferred). If None, uses module-level Config.
    
    Returns:
        Base URL string or None if not found
    """
    config_to_use = config or Config
    configured_url = getattr(config_to_use, "OPENAI_BASE_URL", None)
    if configured_url and configured_url.strip():
        return configured_url.strip()
    return None


def get_openai_key_source(config: Optional[Config] = None) -> Optional[str]:
    """
    Gets the source of the API key from Config.
    
    Args:
        config: Config instance (preferred). If None, checks module-level Config.
    
    Returns:
        Source identifier string or None
    """
    config_to_use = config or Config

    configured_key = _normalize_secret_candidate(getattr(config_to_use, "OPENAI_API_KEY", None))
    if configured_key:
        return "OPENAI_API_KEY"
    return None


def has_valid_api_key(config: Optional[Config] = None) -> bool:
    """
    Checks if a valid API key is configured
    
    Args:
        config: Config instance (preferred). If None, uses module-level Config.
    """
    config_to_use = config or Config
    return resolve_openai_key(config_to_use) is not None


def create_openai_client(options: Optional[ClientOptions] = None, config: Optional[Config] = None) -> Optional[OpenAI]:
    """
    Creates a new OpenAI client with config-based credential resolution
    
    This function follows adapter boundary pattern:
    - Accepts Config instance (preferred) or falls back to env
    - Stateless initialization (no local state dependencies)
    - Deterministic behavior (same inputs = same outputs)
    - Comprehensive error handling and logging
    
    Args:
        options: Client initialization options (optional)
        config: Config instance (preferred). If None, uses module-level Config hydration.
    """
    if options is None:
        options = ClientOptions()
    
    # Use Config if provided, otherwise use module-level Config
    config_to_use = config or Config
    
    trace_id = record_trace_event("openai.client.create.start", {
        "hasApiKeyOverride": bool(options.api_key),
        "hasBaseURLOverride": bool(options.base_url),
        "timeout": options.timeout
    })
    
    try:
        # Resolve API key from config (adapter boundary pattern)
        api_key = options.api_key or resolve_openai_key(config_to_use)
        
        if not api_key:
            logger.warning(
                "OpenAI API key not configured - AI endpoints will return mock responses",
                extra={"operation": "createOpenAIClient", "component": "openai.unified"}
            )
            record_trace_event("openai.client.create.no_key", {"traceId": trace_id})
            return None
        
        # Resolve base URL from config
        base_url = options.base_url or resolve_openai_base_url(config_to_use)
        timeout = options.timeout or API_TIMEOUT_MS_DEFAULT
        
        # Create client instance
        client_kwargs = {
            "api_key": api_key,
            "timeout": timeout / 1000.0  # Convert ms to seconds
        }
        
        if base_url:
            client_kwargs["base_url"] = base_url
        
        client = OpenAI(**client_kwargs)
        
        record_trace_event("openai.client.create.success", {
            "traceId": trace_id,
            "model": config_to_use.OPENAI_MODEL,
            "source": get_openai_key_source(config_to_use)
        })
        
        logger.info(
            "✅ OpenAI client created",
            extra={
                "component": "openai.unified",
                "operation": "createOpenAIClient",
                "model": config_to_use.OPENAI_MODEL,
                "source": get_openai_key_source(config_to_use)
            }
        )
        
        return client
    except Exception as error:
        error_message = str(error)
        logger.error(
            "❌ Failed to create OpenAI client",
            extra={"component": "openai.unified", "operation": "createOpenAIClient"},
            exc_info=error
        )
        
        record_trace_event("openai.client.create.error", {
            "traceId": trace_id,
            "error": error_message
        })
        
        return None


def get_or_create_client(config: Optional[Config] = None) -> Optional[OpenAI]:
    """
    Gets or creates the singleton OpenAI client
    
    Uses singleton pattern for consistent client reuse across the application.
    Initializes on first call with config-based credential resolution.
    
    Args:
        config: Config instance (preferred). If None, uses module-level Config.
    """
    global _singleton_client, _initialization_attempted
    
    if _singleton_client:
        return _singleton_client
    
    # Prevent multiple simultaneous initialization attempts
    if _initialization_attempted:
        logger.warning(
            "OpenAI client initialization already attempted, returning None",
            extra={"component": "openai.unified", "operation": "getOrCreateClient"}
        )
        return None
    
    config_to_use = config or Config
    _initialization_attempted = True
    _singleton_client = create_openai_client(ClientOptions(singleton=True), config=config_to_use)
    
    return _singleton_client


def validate_client_health() -> HealthStatus:
    """
    Validates OpenAI client health
    
    Performs comprehensive health check including:
    - API key configuration
    - Client initialization status
    """
    configured = has_valid_api_key()
    initialized = _singleton_client is not None
    
    health = HealthStatus(
        healthy=configured and initialized,
        api_key_configured=configured,
        api_key_source=get_openai_key_source(Config),
        default_model=Config.OPENAI_MODEL,
        fallback_model=getattr(Config, "FALLBACK_MODEL", "gpt-4"),
    )
    # Record the actual check time for observability
    health.last_check = datetime.now().isoformat()
    if not health.healthy:
        if not configured:
            health.error = "API key not configured"
        elif not initialized:
            health.error = "Client not initialized"
    
    return health


def reset_client() -> None:
    """
    Resets the singleton client
    
    Useful for testing or when credentials change.
    Clears singleton instance and allows re-initialization.
    """
    global _singleton_client, _initialization_attempted
    
    _singleton_client = None
    _initialization_attempted = False
    
    record_trace_event("openai.client.reset", {"component": "openai.unified"})
    logger.info(
        "OpenAI client reset",
        extra={"component": "openai.unified", "operation": "resetClient"}
    )


def get_client() -> Optional[OpenAI]:
    """
    Gets the current singleton client without creating a new one
    
    Returns:
        Current singleton client or None if not initialized
    """
    return _singleton_client


def get_default_model() -> str:
    """Gets the default model from configuration"""
    return Config.OPENAI_MODEL


def get_fallback_model() -> str:
    """Gets the fallback model from configuration"""
    return getattr(Config, "FALLBACK_MODEL", "gpt-4")


def get_gpt5_model(config: Optional[Config] = None) -> str:
    """
    Gets the GPT-5 model from Config.
    
    Args:
        config: Config instance (preferred). If None, uses module-level Config.
    
    Returns:
        GPT-5 model name string
    """
    config_to_use = config or Config
    gpt5_model = getattr(config_to_use, "GPT5_MODEL", None) or getattr(config_to_use, "GPT51_MODEL", None)
    if gpt5_model:
        return gpt5_model
    return "gpt-5.1"


# Exports for backward compatibility
__all__ = [
    "create_openai_client",
    "get_or_create_client",
    "get_client",
    "validate_client_health",
    "reset_client",
    "get_default_model",
    "get_fallback_model",
    "get_gpt5_model",
    "has_valid_api_key",
    "get_openai_key_source",
    "resolve_openai_key",
    "resolve_openai_base_url",
    "API_TIMEOUT_MS_DEFAULT",
    "ClientOptions",
    "HealthStatus"
]
